idn/punycode.go

```diff
@@ -30,6 +31,12 @@ const (
 // This function would return an empty string result for domain names with
 // invalid unicode strings. This function expects domain names in lowercase.
 func ToPunycode(s string) string {
+	// Early check to see if encoding is needed.
+	// This will prevent making heap allocations when not needed.
+	if !needpuny(s) {
+		return strings.ToLower(s)
```

Comment:
[ Quoting <notifications@github.com> in "Re: [dns] Do simple prechecks to av..." ]
>> @@ -30,6 +31,12 @@ const (
>>  // This function would return an empty string result for domain names with
>>  // invalid unicode strings. This function expects domain names in lowercase.
>>  func ToPunycode(s string) string {
>> +	// Early check to see if encoding is needed.
>> +	// This will prevent making heap allocations when not needed.
>> +	if !needpuny(s) {
>> +		return strings.ToLower(s)
>
>When `s` is not lowercased the result will still be lowercased. Without the ToLower the tests fail. I suggest to modify the tests and document this behavior better.

Fair enough.

How can the docs be better?

/Miek

--
Miek Gieben